### DISCUSSION TITLE
Allow custom page paths to be defined in code

### DISCUSSION THREAD

**Discussion by kaedroho** [wagtail/wagtail#7812] [Ideas]
In this discussion, I'd like to propose a new method for customising a page's URL path. This would provide a simpler method to override the URL path up to and including the page's slug (which isn't possible with ``RoutablePageMixin``).

For example, it's common to have to prefix a blog post's slug with the date it was posted on: ``https://mysite.com/blog/2021/02/01/my-blog-post``

Currently, the only supported way is to override the pages ``get_url_parts`` and ``route`` methods. This method is (partially) described in the docs [here](https://docs.wagtail.io/en/v2.12/topics/pages.html#customising-url-patterns-for-a-page-model).

There are a few problems with this method, however:

 - A developer must also override this method for other page classes so that descendant pages get the correct URL
 - It requires overriding two methods, one to tell Wagtail the URL and the other to do the routing
 - It requires specific versions of the page to generate URLs. This could cause performance issues, or be forgotten about

The main goal of implementing the new method would be to make it no longer necessary to fetch specific pages to get a URL. This would improve performance in the admin listings and API, it would also reduce the chance of mistakes caused by forgetting to do this in a plugin or custom code.

## Proposal

I'd like to propose replacing this with a new method by adding a``url_path_component`` field to ``Page``. This will replace the ``slug`` for the purpose of routing to pages. It would be set programmatically rather than editable in the admin and would default to the value of the ``slug`` field.

For example, this is all that would be needed to implement the date in URL customisation for blog posts:

```python
class BlogPostPage(Page):
    date = models.DateField()
    ...

    def get_url_path_component(self):
        # Unlike slug, the URL path component field can contain '/' characters
        return f"{self.date.year}/{self.date.month}/{self.date.day}/{self.slug}"
```

No need to override ``route()``, no need to worry about sub-pages, and the value is saved on the ``Page`` model so the correct URL can be generated without ``.specific``.

### Routing

The route method would be updated to search for sub-pages using the ``url_path_component`` field instead of ``slug``. The difference is we now need to account for the fact that ``url_path_component`` can contain ``'/'`` characters.

This could be done with negligible performance impact* by updating our existing query to search for all the potential options and pick the most specific one.

For example, here's the current routing implementation:

https://github.com/wagtail/wagtail/blob/247030df14cd454cb04db1c39ab2e2ab31ed482c/wagtail/core/models/__init__.py#L816-L825

To account for components that contain a ``/``, we could change this to:

```python
try:
    subpage = self.get_children()
        .filter(
            url_path_component__in=[
                '/'.join(path_components[0:i])
                for i in range(len(path_components))
                path_components
            ]
        )
        .order_by(Length('url_path'))
        .last()
except Page.DoesNotExist:
    raise Http404

num_used_components = len(subpage.url_path_component.split('/'))
remaining_components = path_components[num_used_components:]

return subpage.specific.route(request, remaining_components)
```

\* If we can remove the need for anyone to override ``route()`` we could massively improve performance by removing [this ``.specific`` call](https://github.com/wagtail/wagtail/blob/247030df14cd454cb04db1c39ab2e2ab31ed482c/wagtail/core/models/__init__.py#L825)

**Comment by kaedroho** [comment]
Related comment: https://github.com/wagtail/wagtail/pull/7227#issuecomment-855354798

**Comment by ababic** [comment]
I really like the solution! The proposed routing changes are especially tidy. Just wondering what the implications are for enforcing URL uniqueness? Do we just keep ensuring `slug` uniqueness amongst siblings and not worry about validating the `url_path_component` value at all? I suppose that might make sense if using the `slug` is the stock Wagtail behaviour, and the `url_path_component` field itself isn't editable.

**Reply by kaedroho** [reply]
Thanks!

> Do we just keep ensuring slug uniqueness amongst siblings and not worry about validating the `url_path_component` value at all? 

I think this would be fine. I think clashes are unlikely to happen, and are easy to resolve both in code or by an editor. This is also an issue with the current approach of overriding the ``get_url_parts``/``route`` methods. I think we should definitely mention this limitation in the docs though.

**Reply by ababic** [reply]
Cool! And I guess this value will become part of the `url_path` value for the page and its descendants? There's a conditional bit in `save()` that will likely need tweaking:
```
if not ('update_fields' in kwargs and 'slug' not in kwargs['update_fields']):
```

**Comment by allcaps** [comment]
Nice!

The [extension to the treebeard 'move' method](https://github.com/wagtail/wagtail/blob/247030df14cd454cb04db1c39ab2e2ab31ed482c/wagtail/core/models/__init__.py#L1478)  might need to call `get_url_path_component` to construct the new `url_path`.

**Comment by Nigel2392** [comment]
Any clue if any progress was made on this in the meantime?
